File: common/src/test/java/io/github/notenoughupdates/moulconfig/ChromaColourTest.java
package io.github.notenoughupdates.moulconfig;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.awt.Color;

public class ChromaColourTest {
    void testWithSpeed(Color color, int speed) {
        String stringRepr = speed + ":" + color.getAlpha() + ":" + color.getRed() + ":" + color.getGreen() + ":" + color.getBlue();
        int time = speed > 0 ? (int) (ChromaColour.getSecondsForSpeed(speed) * 1000) : 0;
        ChromaColour fromStatic = ChromaColour.fromRGB(color.getRed(), color.getGreen(), color.getBlue(), time, color.getAlpha());
        if (speed == 0) {
            Assertions.assertEquals(color.getRGB(), fromStatic.getEffectiveColourRGB());
        }
        Assertions.assertEquals(stringRepr, fromStatic.toLegacyString());
        Assertions.assertEquals(fromStatic, ChromaColour.forLegacyString(stringRepr));
    }

    @Disabled("This test takes around ~8 seconds to run, we don't need that overhead")
    @Test
    void testAllColours() {
        for (int i = 0; i <= 0xFFFFFF; i++) {
            Color color = new Color(i);
            testWithSpeed(color, 0);
            testWithSpeed(color, 10);
            testWithSpeed(color, 255);
        }
    }
}

File: common/src/test/java/io/github/notenoughupdates/moulconfig/ChromaColourTest.kt
package io.github.notenoughupdates.moulconfig

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.awt.Color

/**
 * Test to verfiy that [ChromaColour.toLegacyString] and [ChromaColour.forLegacyString] both work as expected, serializing the RGB value.
 */
class ChromaColourTest {
    fun testWithSpeed(jColor: Color, speed: Int) {
        val stringRepr = "$speed:${jColor.alpha}:${jColor.red}:${jColor.green}:${jColor.blue}"
        val time = if (speed > 0) (ChromaColour.getSecondsForSpeed(speed) * 1000).toInt() else 0
        val fromStatic = ChromaColour.fromRGB(jColor.red, jColor.green, jColor.blue, time, jColor.alpha)
        if (speed == 0) // Can't test colour accuracy given that the system time influences this
            Assertions.assertEquals(jColor.rgb, fromStatic.getEffectiveColourRGB())
        Assertions.assertEquals(stringRepr, fromStatic.toLegacyString())
        Assertions.assertEquals(fromStatic, ChromaColour.forLegacyString(stringRepr))
    }

    @Disabled("This test takes around ~8 seconds to run, we don't need that overhead")
    @Test
    fun testAllColours() {
        yieldAllColours()
            .forEach {
                testWithSpeed(it, 0)
                testWithSpeed(it, 10)
                testWithSpeed(it, 255)
            }
    }

    fun yieldAllColours() = (0..0xFFFFFF)
        .asSequence()
        .map { Color(it) }
}
